### DISCUSSION TITLE
The attribute_data collection has empty data both on postman and on custom API response.

### DISCUSSION THREAD

**Discussion by SilenceWF** [lunarphp/lunar#49] [Q&A]
Hi, I love this project specially the v2. I am using the 2.0-beta6 version.

I have a problem though, I've made 2 products just to try things out and when I get all products to my frontend with the the call below:

`Route::get('/products', function (Request $request) {
    return \GetCandy\Models\Product::where('status', 'published')->with('productType', 'variants', 'associations')->get();
});`

The response is this one:

[
    {
        "id": 1,
        "product_type_id": 1,
        "status": "published",
        "attribute_data": {
            "name": {},
            "description": {}
        },
        "brand": "The freaky store",
        "created_at": "2021-12-25T16:07:28.000000Z",
        "updated_at": "2022-01-10T17:28:10.000000Z",
        "deleted_at": null,
        "product_type": {
            "id": 1,
            "name": "Stock",
            "created_at": "2021-12-25T16:00:59.000000Z",
            "updated_at": "2021-12-25T16:00:59.000000Z"
        },
        "variants": [
            {
                "id": 1,
                "product_id": 1,
                "tax_class_id": 1,
                "tax_ref": null,
                "unit_quantity": 10,
                "sku": "900800700",
                "gtin": "900800700",
                "mpn": "900800700",
                "ean": "900800700",
                "length_value": null,
                "length_unit": null,
                "width_value": null,
                "width_unit": null,
                "height_value": null,
                "height_unit": null,
                "weight_value": null,
                "weight_unit": null,
                "volume_value": null,
                "volume_unit": null,
                "shippable": 1,
                "stock": 2,
                "backorder": 0,
                "purchasable": "in_stock",
                "created_at": "2021-12-25T16:07:28.000000Z",
                "updated_at": "2022-01-10T16:34:55.000000Z",
                "deleted_at": null
            }
        ],
        "associations": []
    },
    {
        "id": 2,
        "product_type_id": 1,
        "status": "published",
        "attribute_data": {
            "name": {},
            "description": {}
        },
        "brand": "The freaky store",
        "created_at": "2022-01-10T17:46:58.000000Z",
        "updated_at": "2022-01-10T17:46:58.000000Z",
        "deleted_at": null,
        "product_type": {
            "id": 1,
            "name": "Stock",
            "created_at": "2021-12-25T16:00:59.000000Z",
            "updated_at": "2021-12-25T16:00:59.000000Z"
        },
        "variants": [
            {
                "id": 2,
                "product_id": 2,
                "tax_class_id": 1,
                "tax_ref": null,
                "unit_quantity": 1,
                "sku": "34455668",
                "gtin": "34455668",
                "mpn": "34455668",
                "ean": "34455668",
                "length_value": null,
                "length_unit": null,
                "width_value": null,
                "width_unit": null,
                "height_value": null,
                "height_unit": null,
                "weight_value": null,
                "weight_unit": null,
                "volume_value": null,
                "volume_unit": null,
                "shippable": 1,
                "stock": 3,
                "backorder": 0,
                "purchasable": "in_stock",
                "created_at": "2022-01-10T17:46:58.000000Z",
                "updated_at": "2022-01-10T17:46:58.000000Z",
                "deleted_at": null
            }
        ],
        "associations": []
    }
]

So in "attribute_data" there are no product names or description in any languages. I've set up two languages.
Please help me.

Thank you!

**Comment by glennjacobs** [comment]
Can you see them in the admin hub? How did you create the products, via the admin hub or via code?

**Reply by SilenceWF** [reply]
In the hub the products are all fine. I created the products via the hub after I logged in. I have two currencies and two languages. But in the json the name and description are empty. In database there are not empty neither in hub. I think my api route is wrong I don't know.

**Reply by SilenceWF** [reply]
Or this `'attribute_data' => AsAttributeData::class` doesn't work as expected in the Product model.

**Reply by glennjacobs** [reply]
I think the FieldTypes don't have a `__toString()` function, so it's not able to output it. I'll raise an issue for this. Sorry, I should have kept this as an issue.

However, for API work I would suggest using [API Resources](https://laravel.com/docs/8.x/eloquent-resources) which allows you to set the output format, meaning you can call the `getValue()` method on the fields.

**Reply by SilenceWF** [reply]
It's ok. I will wait for the fix.

Yeah, thank you for the suggestion. I was just playing with the project just to try things out. I hope I helped finding this issue.
I will take a look at **API Resources** now.
Thank you!

**Reply by glennjacobs** [reply]
Strange. I just tested this locally with your code and I get the following...

![image](https://user-images.githubusercontent.com/647407/148822174-68c57a39-9f46-41fb-9a5d-a11a0aeb8a45.png)

So looks like it should work OK...

**Reply by SilenceWF** [reply]
I don't know. I've added that function to FieldType.php file inside Base and I got this:

![Screenshot 2022-01-10 210651](https://user-images.githubusercontent.com/24436254/148824429-cda9d019-4baa-4a42-b6fa-4298b7e20ae6.png)

Did I add the `__toString()` function to the wrong file? 🙃

**Reply by glennjacobs** [reply]
Yeah, that won't work. But I didn't need to add anything in the end, the attribute data shows for me - so I'm a bit confused as to why you're having problems.

@alecritson any ideas?

**Reply by alecritson** [reply]
Could something, maybe a language, not be set as default in the database? Or is the app locale for the app set as a language that doesn't have a value in the db? 🤔

I'll try and pull down the code in the morning and have a squint

**Reply by glennjacobs** [reply]
Looks like @alecritson has worked it out in #56

**Reply by SilenceWF** [reply]
Thank you guys I made the changes and now it is working great! WOOHOO! 🎉

**Comment by glennjacobs** [comment]
I've opened an issue for this #50

**Comment by alecritson** [comment]
There is a PR which should address the issue discussed here #56
